lib/utils/compose.js

```diff
@@ -0,0 +1,1086 @@
+						.pipe(task.logStream);
+				};
+			}
+			return task;
+		})
+		.then(function(tasks) {
+			logger.logDebug('Prepared tasks; building...');
+			return Promise.map(
+				builder.performBuilds(tasks, docker, BALENA_ENGINE_TMP_PATH),
+				function(builtImage) {
+					if (!builtImage.successful) {
+						builtImage.error.serviceName = builtImage.serviceName;
+						throw builtImage.error;
+					}
+
+					const d = imageDescriptorsByServiceName[builtImage.serviceName];
+					const task = _.find(tasks, { serviceName: builtImage.serviceName });
+
+					const image = {
+						serviceName: d.serviceName,
+						name: d.image.tag ?? d.image,
+						logs: truncateString(task.logBuffer.join('\n'), LOG_LENGTH_MAX),
+						props: {
+							dockerfile: builtImage.dockerfile,
+							projectType: builtImage.projectType,
+						},
+					};
+
+					// Times here are timestamps, so test whether they're null
+					// before creating a date out of them, as `new Date(null)`
+					// creates a date representing UNIX time 0.
+					if (builtImage.startTime) {
+						image.props.startTime = new Date(builtImage.startTime);
+					}
+					if (builtImage.endTime) {
+						image.props.endTime = new Date(builtImage.endTime);
+					}
+					return docker
+						.getImage(image.name)
+						.inspect()
+						.get('Size')
+						.then(size => {
+							image.props.size = size;
+						})
+						.return(image);
+				},
+			).tap(function(images) {
+				const summary = _(images)
+					.map(({ serviceName, props }) => [
+						serviceName,
+						`Image size: ${humanize.filesize(props.size)}`,
+					])
+					.fromPairs()
+					.value();
+				renderer.end(summary);
+			});
+		})
+		.finally(renderer.end);
+}
+
+const createRelease = function(apiEndpoint, auth, userId, appId, composition) {
+	const _ = require('lodash');
+	const crypto = require('crypto');
+	const releaseMod = require('balena-release');
+
+	const client = releaseMod.createClient({ apiEndpoint, auth });
+
+	return releaseMod
+		.create({
+			client,
+			user: userId,
+			application: appId,
+			composition,
+			source: 'local',
+			commit: crypto
+				.pseudoRandomBytes(16)
+				.toString('hex')
+				.toLowerCase(),
+		})
+		.then(function({ release, serviceImages }) {
+			return {
+				client,
+				release: _.omit(release, [
+					'created_at',
+					'belongs_to__application',
+					'is_created_by__user',
+					'__metadata',
+				]),
+				serviceImages: _.mapValues(serviceImages, serviceImage =>
+					_.omit(serviceImage, [
+						'created_at',
+						'is_a_build_of__service',
+						'__metadata',
+					]),
+				),
+			};
+		});
+};
+
+const tagServiceImages = (docker, images, serviceImages) =>
+	Promise.map(images, function(d) {
+		const serviceImage = serviceImages[d.serviceName];
+		const imageName = serviceImage.is_stored_at__image_location;
+		const match = /(.*?)\/(.*?)(?::([^/]*))?$/.exec(imageName);
+		if (match == null) {
+			throw new Error(`Could not parse imageName: '${imageName}'`);
+		}
```

Comment:
I added this check to throw a better error in the case it doesn't match rather than an error from trying to destructure null